mmcv/device/ipu/model_converter.py

```diff
@@ -0,0 +1,829 @@
+    def _parse_losses(self, losses):
+        log_vars = OrderedDict()
+        for loss_name, loss_value in losses.items():
+            if isinstance(loss_value, torch.Tensor):
+                log_vars[loss_name] = loss_value.mean()
+            elif isinstance(loss_value, list):
+                log_vars[loss_name] = sum(loss.mean() for loss in loss_value)
+            elif isinstance(loss_value, dict):
+                for name, value in loss_value.items():
+                    log_vars[name] = value
+            else:
+                raise TypeError(
+                    f'{loss_name} is not a tensor or list of tensors')
+
+        loss = sum(value for key, value in log_vars.items()
+                   if 'loss' in key)
+        log_vars['loss'] = loss
+
+        return loss, log_vars
+
+    def forward_eval(self, kwargs):
+        img = kwargs.pop('img')
+        img_metas = kwargs.pop('img_metas', None)
+        return_loss = kwargs.pop('return_loss')
+        assert not return_loss
+        # TODO Temporarily hard-code to close post_process,
+        # otherwise, in the third trace(_check_trace),
+        # post_process will convert output tensor to numpy array automatically,
+        # resulting in _check_trace failure
+        outputs = self.model(
+            img, img_metas=img_metas,
+            return_loss=return_loss, post_process=False)
+        return outputs
+
+
+class PoplarExecutorForMMCV(PoplarExecutor):
+    """An executor for inputs/outputs parsing, model compilation,
+    data alignment and IPU upload/download.
+
+    Args:
+        model (:obj:`nn.Module`): The model to be compiled.
+        logger (:obj:`logging.Logger`): Logger used during running.
+             Defaults to None.
+        training (bool): Model in training mode or eval mode.
+        modules_to_record (mmcv.Config, list): Index or name of modules which
+            will be recorded for output. It is necessary to specify output for
+            static graph of model training or inference.
+        args (argument list): Arguments passed to the `__init__`
+            method of PoplarExecutor.
+        kwargs (keyword arguments): Keyword arguments passed to the `__init__`
+            method of PoplarExecutor.
+    """
+    def __init__(
+            self,
+            model,
+            logger=None,
+            training=True,
+            modules_to_record=None,
+            *args,
+            **kwargs
+            ):
+        # self.model == self._user_model: input pytorch model
+        # self._model: wrapped model which is used to compile
+        # and update weights, these two models use same weights
+        # wrapped model only accept and output tuple, so ComplexDataManager
+        # will convert dictionary to tuple and convert them back
+        self.inputs_tree_manager = ComplexDataManager(logger=logger)
+        self.outputs_tree_manager = ComplexDataManager(logger=logger)
+        self.logger = logger
+        self.hooked_features = {}
```

Comment:
might some explanation of difference between hooked_features_ipu and hooked_features and what they are used for.